cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,484 @@
+				return nil, newFakeNotFoundResponse(), errors.New("unexpected error")
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{*newFakeFirewall(fakeWorkerFWName, fakeInboundRule)}, newFakeOKResponse(), nil
+			},
+			expectedGodoFirewallCreateResp: func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotOKResponse(), errors.New("unexpected error")
+			},
+			expectedError: errFailedToCreateFirewallOnUpdate,
+		},
+		{
+			name:         "return error when failing to add inbound rules on update request to firewall API",
+			fwCache:      newFakeFirewallCache([]godo.InboundRule{diffFakeInboundRule}),
+			inboundRules: []godo.InboundRule{fakeInboundRule},
+			expectedGodoFirewallUpdateResp: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotOKResponse(), errFailedToAddInboundRules
+			},
+			expectedError: errFailedToAddInboundRules,
+		},
+		{
+			name:         "when the firewall cache is nil return existing firewall from API then update cache",
+			fwCache:      newFakeFirewallCacheEmpty(),
+			inboundRules: []godo.InboundRule{fakeInboundRule},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{*newFakeFirewall(fakeWorkerFWName, fakeInboundRule)}, newFakeOKResponse(), nil
+			},
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					listFunc:   test.expectedGodoFirewallListResp,
+					updateFunc: test.expectedGodoFirewallUpdateResp,
+					createFunc: test.expectedGodoFirewallCreateResp,
+					getFunc:    test.expectedGodoFirewallGetResp,
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			err := fc.fwManager.Set(ctx, test.inboundRules)
+			if want, got := test.expectedError, err; want != got {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+		})
+	}
+}
+
+func TestFirewallController_createInboundRules(t *testing.T) {
+	inboundRules := []godo.InboundRule{
+		{
+			Protocol:  "tcp",
+			PortRange: "31220",
+			Sources: &godo.Sources{
+				Tags: []string{"tag"},
+			},
+		},
+		{
+			Protocol:  "tcp",
+			PortRange: "20000",
+			Sources: &godo.Sources{
+				Tags: []string{"tag"},
+			},
+		},
+		{
+			Protocol:  "tcp",
+			PortRange: "40000",
+			Sources: &godo.Sources{
+				Tags: []string{"tag"},
+			},
+		},
+	}
+	testNodePortService := &v1.Service{
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeNodePort,
+			Ports: []v1.ServicePort{
+				{
+					Protocol: "tcp",
+					Port:     31220,
+					NodePort: 31220,
+				},
+			},
+		},
```

Comment:
Could you please add a second `NodePort`-typed Service here and update the test to verify it is included as well? It's always good to test that a function which should process multiple elements really does so successfully.